README.md
freedom-routes, Intenet Freedom - add local routes in vpn for accessing local websites using direct link
========================================================================================================

|                |                                                             |
|----------------|------------------------------------------------------       |
| Homepage:      | https://github.com/GutenYe/freedom-routes        |
| Author:	       | Guten                                            |
| License:       | MIT-LICENSE                                                |
| Documentation: | http://godoc.org/github.com/GutenYe/freedom-routes
| Issue Tracker: | https://github.com/GutenYe/freedom-routes/issues |
| Platforms:     | Linux, Mac OS X, Windows                         |

Use ip data from APNIC to generate routes for vpn to make accessing websites in China use direct link, while accessing foreign websites use vpn. Which make browse websites much faster.

*Please Help*: the mac package can't get into homebrew's main repository for it's un-notable, so if you like the project, please star it at the top of this page.

Getting started
---------------

**Usage**:

	$ freedom-routes [options] <template>
      -o, --output="."                 # output directory

View all avaliable templates at [here](https://github.com/GutenYe/freedom-routes/tree/master/routes/templates)

**Linux (OpenVPN)**

	$ freedom-routes linux -o /etc/openvpn 
	> create /etc/openvpn/routes-up.sh
	> create /etc/openvpn/routes-down.sh

	$ edit /etc/openvpn/x.conf

		script-security 2
		up ./routes-up.sh
		down ./routes-down.sh

**RouterOS**

	$ freedom-routes routeros

* upload freedomroutes.rsc via winbox/WebConsole/ssh/ftp
* get into your router terminal, type `/import freedomroutes.rsc`
* add a default route with `routing-mark` set to `freedomroutes.domestic`
* see `/ip route rule print` for more details..

Others, see documentation at [here](https://github.com/GutenYe/chnroutes/blob/master/README.md) (Chinese)

**Add IPs from local**

	# create /etc/freedom-routes
		70.33.217.1/32

**Auto update IP data**

Ip data may change over time, so it's recommand to update every month.

	$ create /etc/cron.monthly/freedom-routes with 0755 mode

		#!/bin/bash

		freedom-routes linux -o /etc/openvpn

Install
-------

**Use Package**

* ArchLinux: `pacaur -S freedom-routes` # install [freedom-routes](https://aur.archlinux.org/packages/freedom-routes/) from AUR
* Mac OS X: `brew tap gutenye/alt; brew install freedom-routes`
* Windows: download [x64](http://downloads.gutenye.com/freedom-routes/freedom-routes.windows.amd64.zip) [x86](http://downloads.gutenye.com/freedom-routes/freedom-routes.windows.386.zip)

**Use Source**

	$ [sudo] go get github.com/GutenYe/freedom-routes
	
**Dependencies**

* iproute2 (Linux, Mac)

Development 
===========

Build
------

	mkdir output
	sed -i '/const ASSETS_MODE/s/.*/const ASSETS_MODE = "runtime"/' routes/routes.go
	go build -o output/freedom-routes
	cp -r routes/templates output

Contributing 
-------------

* Submit any bugs/features/ideas to github issue tracker.

Please see [Contibution Documentation](https://github.com/GutenYe/freedom-routes/blob/master/CONTRIBUTING.md).

A list of [Contributors](https://github.com/GutenYe/freedom-routes/contributors).

Resources
---------

* [chnroutes](https://github.com/GutenYe/chnroutes): scripts help chinese netizen with vpn.
* [Pacaur](https://github.com/Spyhawk/pacaur): An AUR helper for ArchLinux.
* [Homebrew](https://github.com/mxcl/homebrew): The missing package manager for Mac OS X.

Copyright
---------

(the MIT License)

Copyright (c) 2013 Guten

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## 1.0.0 (xx)

First version.


build.sh
#!/bin/bash

# Build binary for distribution.
#
# Usage:
#
#   ./build.sh [options] [platform ...]
#     -d <assets_dir>
#     -p              # packaging the binary
#
# Example:
#
#   ./build.sh -d /usr/share/freedom-routes    # build with system go
#   ./build.sh release                         # build $RELEASEs with user complied go and upload the package to s3.

VERSION=$(sed -rn 's/.*const VERSION.*"([0-9.]+)".*/\1/p' main.go)
FILES="routes/templates freedom-routes.etc README.md CHANGELOG.md"
RELEASE="homebrew/amd64 homebrew/386 windows/386 windows/amd64"
declare -A OS_MAP=(
	[homebrew]="darwin"
)
declare -A DIR_MAP=(
	[homebrew]="/usr/share/freedom-routes"
)

platform=""
os=""
arch=""
assets_dir=""

function cgo_enabled { [ "$1" == "$GOHOSTOS" ] && echo 1 || echo 0; }
function ext { [ $1 == "windows" ] && echo .exe || echo ""; }

# dist{platform, os, arch, assets_dir}
function dist {
	rm -r dist 2>/dev/null
	mkdir dist

	cp -r $FILES dist
	build
}

# build{platform, os, arch, assets_dir}
function build {
	echo -e "\nbuilding $platform/$arch"

	sed -i "/const ASSETS_MODE/s~.*~const ASSETS_MODE = \"$assets_dir\"~" routes/routes.go
	CGO_ENABLED=$(cgo_enabled $os) GOOS=$os GOARCH=$arch $GOROOT/bin/go build -o "dist/freedom-routes$(ext $os)"
	sed -i '/const ASSETS_MODE/s/.*/const ASSETS_MODE = "source"/' routes/routes.go
}

# package{platform, os, arch}
function package {
	echo "packing $platform/$arch" 
	mkdir dist/freedom-routes-$VERSION
	cp -r dist/* dist/freedom-routes-$VERSION 2>/dev/null

	case $os in
		linux | darwin )
			tar zcvf freedom-routes.$platform.$arch-$VERSION.tar.gz -C dist freedom-routes-$VERSION;;
		windows ) 
			rm ../freedom-routes.$platform.$arch.zip 2>/dev/null
			cd dist && zip -r ../freedom-routes.$platform.$arch.zip freedom-routes-$VERSION && cd ..
			;;
	esac
}

function upload {
	for file in $*; do
		s3cmd put --acl-public $file s3://downloads.gutenye.com/freedom-routes/
		#current=$(echo $file | sed -r 's/(.*)-[0-9.]+(\..*)$/\1\2/')
		#s3cmd copy s3://downloads.gutenye.com/freedom-routes/$file s3://downloads.gutenye.com/freedom-routes/$current
	done
}

#
# ¤main
# ----

eval "$(go env)"

o_assets_dir="runtime"
o_package=false
while getopts "d:p" opt; do
  case $opt in
    d ) o_assets_dir=$OPTARG ;;
		p ) o_package=true ;;
  esac
done
shift $(( OPTIND - 1 ))

case $1 in
	"" )
		platform=$GOOS; os=$GOOS; arch=$GOARCH; assets_dir=$o_assets_dir
		dist
		[ $o_package == true ] && package
		;;
	release )
		GOROOT="/home/guten/dev/src/go/go"
		rm *.zip *.tar.gz 2>/dev/null
		for release in $RELEASE; do
			platform=${release%/*}; arch=${release#*/}; os=${OS_MAP[$platform]-$platform}
			assets_dir=${DIR_MAP[$platform]-$o_assets_dir}
			dist
			package
		done
		upload *.zip *.tar.gz
		;;
	upload )
		shift
		upload $*
		;;
	* )
		GOROOT="/home/guten/dev/src/go/go"
		platform=${1%/*}; arch=${1#*/}; os=${OS_MAP[$platform]-$platform}
		dist
		[ $o_package == true ] && package
		;;
esac

exit 0


routes/routes.go
// Package routes is the library for freedom-routes.
package routes

import (
  "fmt"
  "strings"
  "regexp"
  "strconv"
  "os"
  "io/ioutil"
  "path/filepath"
  "net"
  "net/http"
  "math"
  "text/template"
  "runtime"
)

const ASSETS_MODE = "source"
var ASSETS_PATH = getAssetsPath(ASSETS_MODE)
var REMOTE_URL = "http://ftp.apnic.net/apnic/stats/apnic/delegated-apnic-latest"
var LOCAL_PATH = "/etc/freedom-routes"

type Ip struct {
  Ip string
  Cidr string
  Mask string
}

func (ip Ip) String() string {
  return fmt.Sprintf("%s/%s", ip.Ip, ip.Cidr)
}

// Generate result files from template.
func Generate(templateName string, ips []Ip, outputDir string) {
  templateDir := filepath.Join(ASSETS_PATH, "templates", templateName)

  type Data struct {
    Ips []Ip
  }
  data := Data{ips}

	files, err := ioutil.ReadDir(templateDir)
  if err != nil { panic(err) }
	for _, file := range files {
    templatePath := filepath.Join(templateDir, file.Name())
    output := filepath.Join(outputDir, file.Name())
    
    tmpl, err := template.ParseFiles(templatePath)
    if err != nil { panic(err) }
    file, err := os.Create(output)
    if err != nil { panic(err) }
    defer file.Close()
    if runtime.GOOS != "windows" {
      err = file.Chmod(0755)
      if err != nil { panic(err) }
    }
    err = tmpl.Execute(file, data)
    if err != nil { panic(err) }
    fmt.Printf("Create %s\n", output)
  }
}

// Fetch local ips and remote ips.
func FetchIps() (ips []Ip) {
  ips = FetchLocalIps()
  ips = append(ips, FetchRemoteIps()...)

  return ips
}

// Fetch remote ips from REMOTE_URL.
//
// syntax:
//
//   apnic|CN|ipv4|1.94.0.0|131072|20100806|allocated
//   ...
func FetchRemoteIps() (ips []Ip) {
  println("Fetching latest ip data from apnic.net, this may take a few minutes, please wait...")

  resp, err := http.Get(REMOTE_URL)
  if err != nil { panic(err) }

  defer resp.Body.Close()
  body, err := ioutil.ReadAll(resp.Body)
  if err != nil { panic(err) }

  re, _ := regexp.Compile(`apnic\|CN\|ipv4\|([\d\.]+)\|(\d+)\|`)
  rows := re.FindAllSubmatch(body, -1)
  for i := 0; i < len(rows); i++ {
    row := rows[i]
    ip := string(row[1])
    num_ip, _ := strconv.Atoi(string(row[2]))
    cidr := 32 - int(math.Log2(float64(num_ip)))
    cidrStr := strconv.Itoa(cidr)

    ips = append(ips, Ip{ip, cidrStr, cidr2mask(cidr)})
  }

  return ips
}


// Fetch local ips from LOCAL_PATH.
//
// syntax:
//
//   70.33.217.25/32
//   ...
func FetchLocalIps() (ips []Ip) {
  data, err := ioutil.ReadFile(LOCAL_PATH)
  if os.IsNotExist(err) {
    return ips
  } else if err != nil {
    panic(err)
  }

  lines := strings.Split(string(data), "\n")
  for _, line := range lines {
    line = strings.Trim(line, " \n\t")
    if line == "" { continue }
    tokens := strings.Split(line, "/")
    ip := tokens[0]
    cidr := tokens[1]
    cidrInt, _ := strconv.Atoi(cidr)
    ips = append(ips, Ip{ip, cidr, cidr2mask(cidrInt)})
  }

  return ips
}

// Convert CIDR to Mask.
// 
// Example:
//
//   cidr2mask(24)       #-> "255.255.255.0"
func cidr2mask(cidr int) string {
  mask := net.CIDRMask(cidr, 32)
  masks := []string{}
  for _, v := range mask {
    masks = append(masks, strconv.Itoa(int(v))) 
  }

  return strings.Join(masks, ".")
}
